Add uint32/uint64 support for tf.tile

yongtang opened the pull request:
This PR tries to address the issue raised in #39405 where there is no uint32/uint64 support for tf.tile.

The related kernel impl for uint32 and uint64 has been added in this PR.

This PR fixes #39405

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for uint32/uint64 support of tf.tile

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add uint32/uint64 support for tf.tile

This PR tries to address the issue raised in 39405 where
there is no uint32/uint64 support for tf.tile.

The related kernel impl for uint32 and uint64 has been added in this PR.

This PR fixes 39405

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Bazel buildifier lint fix

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

ezhulenev commented:
This PR breaks `tensorflow/lite/experimental/delegates/coreml:coreml_diff_test_IOS_12`

```
Undefined symbols for architecture x86_64:
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned int, int>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<int const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, int>::HandleCaseImpl<(tensorflow::DataType)22>(tensorflow::OpKernelContext*, absl::Span<int const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned int, long long>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<long long const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, long long>::HandleCaseImpl<(tensorflow::DataType)22>(tensorflow::OpKernelContext*, absl::Span<long long const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned long long, int>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<int const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, int>::HandleCaseImpl<(tensorflow::DataType)23>(tensorflow::OpKernelContext*, absl::Span<int const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned long long, long long>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<long long const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, long long>::HandleCaseImpl<(tensorflow::DataType)23>(tensorflow::OpKernelContext*, absl::Span<long long const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
ld: symbol(s) not found for architecture x86_64
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

I'll take a look at this build errors next week, if you have time to find what's going on before that I'd appreciate that. From the error message it seems that build error is reproducible with a regular x86 compiler.

yongtang committed:
Add tile_functor_cpu_uint[32|64].cc to android_extended_ops_group2 BAZEL rule

tile_functor_cpu_uint[32|64].cc are files that were added to tile_ops
in BUILD. However, this might also be needed in android_extended_ops_group2
(like other similiar functor cc sources). This commit adds
tile_functor_cpu_uint[32|64].cc to android_extended_ops_group2 BAZEL rule
just in case they could solve the following issues:
```
Undefined symbols for architecture x86_64:
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned int, int>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<int const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, int>::HandleCaseImpl<(tensorflow::DataType)22>(tensorflow::OpKernelContext*, absl::Span<int const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned int, long long>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<long long const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, long long>::HandleCaseImpl<(tensorflow::DataType)22>(tensorflow::OpKernelContext*, absl::Span<long long const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned long long, int>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<int const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, int>::HandleCaseImpl<(tensorflow::DataType)23>(tensorflow::OpKernelContext*, absl::Span<int const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
  "tensorflow::functor::Tile<Eigen::ThreadPoolDevice, unsigned long long, long long>::operator()(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, absl::Span<long long const>) const", referenced from:
      void tensorflow::TileOp<Eigen::ThreadPoolDevice, long long>::HandleCaseImpl<(tensorflow::DataType)23>(tensorflow::OpKernelContext*, absl::Span<long long const> const&, tensorflow::Tensor*) in libportable_tensorflow_kernels.lo(tile_ops_8df9402447fb5c99fae84bd69e6311f0.o)
ld: symbol(s) not found for architecture x86_64
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @ezhulenev for the help

 I don't have access to ios build. However, this PR adds two additional new files (for tile_ops functor template specialization). I also noticed similar tile_ops functor c++ files are added both in `tile_ops` rule and `android_extended_ops_group2` rule of the `tensorflow/core/kernels/BUILD` file. My wild guess is that `android_extended_ops_group2` was used for ios as well, thus the ios `Undefined symbols` error case if those two files are not in `android_extended_ops_group2`.

I updated the PR with two functor files added to `android_extended_ops_group2` to give it a try. Maybe you can take a look and see if the issue could be resolved? If not I will try to see if I can find something else.

ezhulenev commented:
Thanks! Will try to pull it in if tooling will work, otherwise will do manual pull, unlikely earlier than tuesday.
